book/optimization.qmd

```diff
@@ -1,173 +1,236 @@
-#| echo: false
+{{< include _setup.qmd >}}
 
-knitr::include_graphics("images/hpo.png")
+Machine learning algorithms usually include parameters and hyperparameters.
+Parameters are what we might think of as model coefficients or weights, when fitting a model we are essentially just running algorithms that fit parameters.
+In contrast, hyperparameters, are configured by the user and determine how the model will fit its parameters.
+Examples include setting the number of trees in a random forest, penalty variables in SVMs, or the learning rate in a neural network.
+Building a neural network is sometimes referred to as an 'art' as there are so many hyperparameters to configure that strongly influence model performance, this is also true for other machine learning algorithms.
+So in this chapter, we will demonstrate how to make this into more of a science.
+
+The goal of hyperparameter optimization (HPO) (@sec-model-tuning) or model tuning is to find the optimal configuration of hyperparameters of an ML algorithm for a given task.
+There is no closed-form mathematical representation (nor analytic gradient information) for model agnostic HPO, instead we follow a numerical black-box optimization: an ML algorithm is configured with values chosen for one or more hyperparameters, this algorithm is then evaluated (optimally with a robust resampling method) and its performance measured, this is repeated with multiple configurations and the configuration with the best performance is selected. We could think of finding the optimal configuration in the same way as selecting a model from a benchmark experiment, where in this case each model uses the same underlying algorithm but with different hyperparameter configurations. For example we could naively tune the number of trees in a random forest using basic mlr3 code:
+
+```{r}
+#| fig-cap: In this code example we benchmark three random forest models with 1, 10, and 100 trees respectively, using 3-fold resampling, classification error loss, and tested on the simplified penguins dataset. The plot shows that the models with 10 and 100 trees are better performing across all three folds and 100 trees may be better than 10.
+#| fig-alt: Boxplots for each of the three configurations showing classification error over the three folds. Image shows worst performance in model with 1 trees and similar performance with 10 and 100 trees.
+autoplot(benchmark(benchmark_grid(
+  tsk("penguins_simple"),
+  c(lrn("classif.ranger", num.trees = 1, id = "1 tree"),
+    lrn("classif.ranger", num.trees = 10, id = "10 trees"),
+    lrn("classif.ranger", num.trees = 100, id = "100 trees")),
+  rsmp("cv", folds = 3)
+)))
+```
+
+Human trial-and-error (which is essentially what we are doing above), is time-consuming, often biased, error-prone, and computationally irreproducible.
+Instead, many sophisticated HPO methods (@sec-tuner) (or 'tuners') have been developed over the last few decades for robust and efficient HPO.
```

Comment:
This feels a bit out of place, needs a better connection. ALso maybe explain what evolutionary algs or Bayesian methods are? A couple examples